### ISSUE TITLE
invalid use of incomplete type ‘struct std::hash<faiss::gpu::AllocType>’

### ISSUE THREAD

**Issue by aptxna** [facebookresearch/faiss#1506]
Hi there!

I was installing FAISS on a redhat-linux server following the instructions on faiss github-page, and encountered these issues during "make" build process.
```
[ 35%] Building CXX object faiss/CMakeFiles/faiss.dir/gpu/StandardGpuResources.cpp.o
In file included from /usr/include/c++/4.8.2/bits/hashtable.h:35:0,
                 from /usr/include/c++/4.8.2/unordered_map:47,
                 from /u/faiss/faiss/gpu/utils/StackDeviceMemory.h:15,
                 from /u/faiss/faiss/gpu/StandardGpuResources.h:12,
                 from /u/faiss/faiss/gpu/StandardGpuResources.cpp:9:
/usr/include/c++/4.8.2/bits/hashtable_policy.h: In instantiation of ‘struct std::__detail::_Hash_code_base<faiss::gpu::AllocType, std::pair<const faiss::gpu::AllocType, std::pair<int, long unsigned int> >, std::__detail::_Select1st, std::hash<faiss::gpu::AllocType>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, true>’:
/usr/include/c++/4.8.2/bits/hashtable_policy.h:1402:10:   required from ‘struct std::__detail::_Hashtable_base<faiss::gpu::AllocType, std::pair<const faiss::gpu::AllocType, std::pair<int, long unsigned int> >, std::__detail::_Select1st, std::equal_to<faiss::gpu::AllocType>, std::hash<faiss::gpu::AllocType>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Hashtable_traits<true, false, true> >’
/usr/include/c++/4.8.2/bits/hashtable.h:174:11:   required from ‘class std::_Hashtable<faiss::gpu::AllocType, std::pair<const faiss::gpu::AllocType, std::pair<int, long unsigned int> >, std::allocator<std::pair<const faiss::gpu::AllocType, std::pair<int, long unsigned int> > >, std::__detail::_Select1st, std::equal_to<faiss::gpu::AllocType>, std::hash<faiss::gpu::AllocType>, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >’
/usr/include/c++/4.8.2/bits/unordered_map.h:100:18:   required from ‘class std::unordered_map<faiss::gpu::AllocType, std::pair<int, long unsigned int> >’
/u/faiss/faiss/gpu/StandardGpuResources.cpp:38:57:   required from here
/usr/include/c++/4.8.2/bits/hashtable_policy.h:1070:12: error: invalid use of incomplete type ‘struct std::hash<faiss::gpu::AllocType>’
     struct _Hash_code_base<_Key, _Value, _ExtractKey, _H1, _H2,

In file included from /usr/include/c++/4.8.2/bits/basic_string.h:3033:0,
                 from /usr/include/c++/4.8.2/string:52,
                 from /u/faiss/faiss/impl/FaissException.h:14,
                 from /u/faiss/faiss/impl/FaissAssert.h:13,
                 from /u/faiss/faiss/gpu/GpuResources.h:11,
                 from /u/faiss/faiss/gpu/StandardGpuResources.h:11,
                 from /u/faiss/faiss/gpu/StandardGpuResources.cpp:9:
/usr/include/c++/4.8.2/bits/functional_hash.h:58:12: error: declaration of ‘struct std::hash<faiss::gpu::AllocType>’
     struct hash;
.
.
.
```
mainly about the errors: " invalid use of incomplete type ‘struct std::hash<faiss::gpu::AllocType>’ " and " declaration of ‘struct std::hash<faiss::gpu::AllocType>’ "

Can someone help?

Following's my cmake process: (cmake version - 3.18.4)
```
-- The CXX compiler identification is GNU 4.8.5
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The CUDA compiler identification is NVIDIA 9.0.176
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Found OpenMP_CXX: -fopenmp (found version "3.1") 
-- Found OpenMP: TRUE (found version "3.1")  
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Could NOT find MKL (missing: MKL_LIBRARIES) 
-- Looking for sgemm_
-- Looking for sgemm_ - not found
-- Looking for sgemm_
-- Looking for sgemm_ - found
-- Found BLAS: /usr/lib64/libopenblas.so  
-- Looking for cheev_
-- Looking for cheev_ - found
-- A library with LAPACK API found.
-- Found CUDAToolkit: /usr/local/cuda/include (found version "9.0.176") 
-- Found SWIG: /usr/bin/swig (found version "2.0.10") found components: python 
-- Found Python: /group/anaconda3/envs/py377/include/python3.7m (found version "3.7.7") found components: Development NumPy Interpreter Development.Module Development.Embed 
-- The C compiler identification is GNU 4.8.5
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Found PythonInterp: /group/anaconda3/envs/py377/bin/python (found version "3.7.7") 
-- Configuring done
CMake Warning (dev) in faiss/CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "faiss".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in faiss/gpu/test/CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "TestGpuDistance".
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) in faiss/gpu/test/CMakeLists.txt:
  Policy CMP0104 is not set: CMAKE_CUDA_ARCHITECTURES now detected for NVCC,
  empty CUDA_ARCHITECTURES not allowed.  Run "cmake --help-policy CMP0104"
  for policy details.  Use the cmake_policy command to set the policy and
  suppress this warning.

  CUDA_ARCHITECTURES is empty for target "TestGpuSelect".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /u/faiss/build
```

**Comment by mdouze** [comment]
Maybe the compiler is a bit old?

**Comment by Aurevoir-68** [comment]
@aptxna    Hello, I have the same problem as you. How can I solve it? Upgrade Compiler?

**Comment by Aurevoir-68** [comment]
@ mdouze

**Comment by aptxna** [comment]
> @aptxna Hello, I have the same problem as you. How can I solve it? Upgrade Compiler?

I used "pip install ...", maybe my compiler or cmake was too old to process the FAISS installing from source code, and I failed to update it because of my root access restriction, so I just gave up and pip install...

**Comment by wickedfoo** [comment]
This appears to be a GCC libstdc++ bug fixed in 6.1. Upgrading gcc should fix it.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60970

**Comment by Anytious** [comment]
I have the same question when compile faiss on gcc5.4 , so sad.
